src/isl.rs

```diff
@@ -0,0 +1,211 @@
+                Some(name) => name,
+                None => {
+                    return Err(invalid_schema_error_raw(
+                        "A type name symbol token does not have any text",
+                    ))
+                }
+            };
+            // TODO: add more constraints to match below
+            let constraint = match constraint_name {
+                "all_of" => {
+                    if value.ion_type() != IonType::List {
+                        return Err(invalid_schema_error_raw(format!(
+                            "all_of constraint was a {:?} instead of a list",
+                            value.ion_type()
+                        )));
+                    }
+                    let types: Vec<IslTypeRef> = value
+                        .as_sequence()
+                        .unwrap()
+                        .iter()
+                        .map(|e| IslTypeRef::parse_from_ion_element(e))
+                        .collect::<IonSchemaResult<Vec<IslTypeRef>>>()?;
+                    IslConstraint::AllOf(types)
+                }
+                "type" => {
+                    if value.ion_type() != IonType::Symbol && value.ion_type() != IonType::Struct {
+                        return Err(invalid_schema_error_raw(format!(
+                            "type constraint was a {:?} instead of a symbol/struct",
+                            value.ion_type()
+                        )));
+                    }
+                    let type_reference: IslTypeRef = IslTypeRef::parse_from_ion_element(value)?;
+                    IslConstraint::Type(type_reference)
+                }
+                _ => {
+                    return Err(invalid_schema_error_raw(
+                        "Type: ".to_owned()
+                            + &type_name.unwrap()
+                            + " can not be built as constraint: "
+                            + constraint_name
+                            + " does not exist",
+                    ))
+                }
+            };
+            constraints.push(constraint);
+        }
+        Ok(IslType::new(type_name.to_owned(), constraints))
+    }
+}
+
+/// Provides an internal representation of schema type reference.
+/// Type reference grammar is defined in [Ion Schema Spec]
+/// [Ion Schema spec]: https://amzn.github.io/ion-schema/docs/spec.html#grammar
+#[derive(Debug, Clone)]
+pub enum IslTypeRef {
+    /// represents core ion type reference
+    IslCoreType(IonType),
+    /// represents a type reference which represents a type imported from another schema
+    AliasType(String),
+    /// represents a type reference defined as an inlined import type from another schema
+    // TODO: add ImportType(Import) where ImportType could either point to a schema represented by an id with all the types or a single type from inside it
+    /// represents an unnamed type definition reference
+    AnonymousType(IslType),
```

Comment:
Okay, now I see that you do have an AnonymousType.

question – Rather than using the `AnonymousType` as a wrapper for `IslType` (which has an optional name), do you think it would make sense to make `AnonymousType` just have its own list of constraints, and use a trait or something for what `AnonymousType` and `IslType` have in common (the getter for the constraints, I guess)?

Reply:
I think this is similar to having two different implementation of `TypeDefinition` but rather for the upper layer `IslType`. I will add it in the issue: #25.